src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -71,51 +87,659 @@ class IntrinsicCall(Call):
-    #: argument is described by an entry in the Dict.
-    _optional_args = {}
-    # The positional arguments to allocate must all be References (or
-    # ArrayReferences but they are a subclass of Reference).
-    _required_args[Intrinsic.ALLOCATE] = ArgDesc(1, None, Reference)
-    _optional_args[Intrinsic.ALLOCATE] = {
-        # Argument used to specify the shape of the object being allocated.
-        "mold": Reference,
-        # Argument specifying both shape and initial value(s) for the object
-        # being allocated.
-        "source": Reference,
-        # Integer variable given status value upon exit.
-        "stat": Reference,
-        # Variable in which message is stored upon error. (Requires that 'stat'
-        # also be provided otherwise the program will just abort upon error.)
-        "errmsg": Reference}
-    _required_args[Intrinsic.DEALLOCATE] = ArgDesc(1, None, Reference)
-    _optional_args[Intrinsic.DEALLOCATE] = {"stat": Reference}
-    _required_args[Intrinsic.RANDOM_NUMBER] = ArgDesc(1, 1, Reference)
-    _optional_args[Intrinsic.RANDOM_NUMBER] = {}
-    _required_args[Intrinsic.MINVAL] = ArgDesc(1, 1, DataNode)
-    _optional_args[Intrinsic.MINVAL] = {
-        "dim": DataNode, "mask": DataNode}
-    _required_args[Intrinsic.MAXVAL] = ArgDesc(1, 1, DataNode)
-    _optional_args[Intrinsic.MAXVAL] = {
-        "dim": DataNode, "mask": DataNode}
-    _required_args[Intrinsic.SUM] = ArgDesc(1, 1, DataNode)
-    _optional_args[Intrinsic.SUM] = {
-        "dim": DataNode, "mask": DataNode}
-    _required_args[Intrinsic.TINY] = ArgDesc(1, 1, (Reference, Literal))
-    _optional_args[Intrinsic.TINY] = {}
-    _required_args[Intrinsic.HUGE] = ArgDesc(1, 1, (Reference, Literal))
-    _optional_args[Intrinsic.HUGE] = {}
+    class Intrinsic(IAttr, Enum):
+        ''' Enum of all intrinsics with their attributes as values using the
+        IAttr namedtuple format:
+
+            NAME = IAttr(name, is_pure, is_elemental, is_inquiry,
+                         required_args, optional_args)
+
+        Note that name is duplicated inside IAttr because each item in the
+        Enum must have a different value, and without the name that would
+        not be guaranteed.
+
+        '''
+        # Fortran special-case statements (technically not Fortran intrinsics
+        # but in PSyIR they are represented as Intrinsics)
+        ALLOCATE = IAttr(
+            'ALLOCATE', False, False, False,
+            ArgDesc(1, None, Reference),
+            {"mold": Reference, "source": Reference, "stat": Reference,
+             "errmsg": Reference})
+        DEALLOCATE = IAttr(
+            'DEALLOCATE', False, False, False,
+            ArgDesc(1, None, Reference), {"stat": Reference})
+        NULLIFY = IAttr(
+            'NULLIFY', False, False, False,
+            ArgDesc(1, None, Reference), {})
+
+        # Fortran Intrinsics (from Fortran 2018 standard table 16.1)
```

Comment:
I don't think we should reuse the fparser as this would create a dependency from PSyIR core to fparser (skipping the reader) and this should be independent of the frontend we use to parse.

Also the pdf parsing is tricky as in some cases (but not in other) the intrinsic are defined by multiple entries. And some thing like ALLOCATE are not even in that table.